### DISCUSSION TITLE
(NEW SCRIPT) IntelliFlow: Intelligent ('context-aware') menu acting like a services/command provider for  Freeplane

### DISCUSSION THREAD

**Discussion by i-plasm** [freeplane/freeplane#1534] [Show and tell]
# IntelliFlow: Intelligent ('context-aware') menu acting like a services/command provider for  Freeplane

Hello dear community,

I'm excited to introduce the [IntelliFlow prototype](https://github.com/i-plasm/freeplane-scripts/wiki/IntelliFlow-Updates). It is an intelligent ('context-aware') menu acting like a services/command provider that aims to make Freeplane workflows faster and more flexible.

The user launches IntelliFlow, and a popup menu appears with services adapted to the context. Both nodes (__'NODE' MODE__) and text components (__'TEXT' mode__) can serve as inputs to the services. Detection is automatic.

![image-intelliflow](https://github.com/freeplane/freeplane/assets/144830522/9a37817e-6790-4152-b78a-3a7bbf8df2a1)

## Intelligent 'Plugin of plugins' / Services aggregator
I conceive this tool as an intelligent  "___plugin of plugins___". With it one can provide services according to the 'mode':

- __'NODE' MODE__. Acts on node (and thus on maps and freeplane-wide). Examples of services/scripts would be those for filtering maps, for executing operation on node or its properties, and in reality, anything the Freeplane API allows.
- __'TEXT' MODE__. Acts on text or portions of text. Example of services would be browsing, searching, transforming or using portions of clicked/selected text
  - ___KEY FEATURE: 'TEXT' MODE CAN ACT ON ANY EDITOR/TEXT COMPONENT___: Markdown (viewer or editor), Html editor, in-line node core editor, notes editor. NOTE: they can't be external editors, unless you are on a Mac.

### Video presentation

This video briefly illustrates the basic concepts:

https://github.com/freeplane/freeplane/assets/144830522/79acfd67-382b-484b-949d-504c85fda60a

## Context Awareness

It is 'Context-Aware' because the mode is determined automatically.

It is also 'Context-Aware' because services can be designed to appear in _both modes, in one, or none_, in a fixed way, or according to prevalidation of conditions.

## Default Services

**NOTE: none of the current default services modify the map content in any way, except the Sketcher service when it adds a blank image if the node has no image.**

### (Default service) SmartOpen

Smart and easy detection and browsing of resources: file paths, app/web URIs, and Freeplane nodes.

- 'NODE' mode: if the node has a link property, it will browse/open it.
- 'TEXT' mode: the clicked or selected text will be opened (if it exists). It can be anything (except executables): a website, a URI, a file, a node ID.
    - Ease of use. In general, you don't need to select your target. Just click anywhere on it, and it will be detected. Exception: if your target is plain text containing whitespaces, you need to select it.
    - Intelligent detection. It can usually detect path/uri/nodeID even if it is not an html hyperlink, but just plain text
    - Files: It can process both relative and absolute paths.
    - URIs: It can process URIs, both relative and absolute.
    - Jump to nodes: It understands node IDs and "jumps" to the node.

### (Default service) Sketcher

- 'NODE' mode: Quickly sketch drawings on a node. Select a node, call the service and a new blank image or an existing one will be opened in a external image editor. Changes to the image will be refreshed in your mindmap. 
- More information on Sketcher: https://github.com/freeplane/freeplane/discussions/1496

### (Default service) Web Search

- 'NODE' mode: search the node core text in the web
- 'TEXT' mode: search the text selection in the web 

### (Default service) Breadcrumb node

- 'NODE' mode: Filter the selected node so that both its ancestors (the "breadcrumb") and its descendants are visible. The descendants preserve the original folding state.

## Permissions

The following script permissions are required (set them in Preferences -> Plug-ins): __1- Permit File read; 2- Permit File write; 3- Permit to execute other applications__

## (Important!) Windows and Linux require a shortcut

... otherwise 'TEXT' mode will ***not*** be available. Go to Freeplane's ` Tools -> Assign hot key `.

## Get IntelliFlow

Script location: ` i-plasm -> Misc Utils -> Prototype For Intelliflow `

[Get IntelliFlow here.](https://github.com/i-plasm/freeplane-scripts/wiki/IntelliFlow-Updates)

---

It is my hope this tool will turn to be useful in many contexts. It has been already quite useful to me! 

It is a prototype, and if it goes well and conditions permit, I'll improve it and make it more capable and customizable. Getting the UI interaction with the different Freeplane elements/focus systems was a bit challenging, but I think I finally got the foundational interaction right (at least on my tests on a Windows, a Mac and a Linux machine). Still, if you observe any area that misbehaves or there is room for improvement please let me know. Your comments are welcome!

**Comment by nohaboat** [comment]
Whenever I try to open external links (red arrow links) or try to websearch using google/bing the external application (eg. browser/file explorer) doesn't open. I am using freeplane version 1.13.1 on archlinux.

When the Intelliflow option to search web or open link is clicked the terminal shows the following error output -

Feb 15, 2026 10:06:34 AM org.freeplane.core.util.LogUtils severe
SEVERE: Exception in thread "FreeplaneUtils: OPEN_RESOURCE"
java.lang.NullPointerException
at java.base/java.lang.ProcessBuilder.start(Unknown Source)
at java.base/java.lang.ProcessBuilder.start(Unknown Source)
at java.base/java.lang.Runtime.exec(Unknown Source)
at java.base/java.lang.Runtime.exec(Unknown Source)
at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
at scripts.IntelliFlow$IOUtils.openViaOSCommand(prototypeForIntelliflow.groovy:2352)
at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
at scripts.IntelliFlow$FreeplaneTools.altOpenOtherOS(prototypeForIntelliflow.groovy:2009)
at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
at java.base/java.lang.reflect.Method.invoke(Unknown Source)
at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
at groovy.lang.MetaClassImpl.invokeStaticMethod(MetaClassImpl.java:1586)
at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:611)
at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.invokeMethodN(ScriptBytecodeAdapter.java:183)
at scripts.IntelliFlow$FreeplaneTools$2.methodMissing(prototypeForIntelliflow.groovy)
at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
at java.base/java.lang.reflect.Method.invoke(Unknown Source)
at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
at groovy.lang.MetaClassImpl.invokeMissingMethod(MetaClassImpl.java:924)
at groovy.lang.MetaClassImpl.invokePropertyOrMissing(MetaClassImpl.java:1413)
at groovy.lang.MetaClassImpl.doInvokeMethod(MetaClassImpl.java:1335)
at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1088)
at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1007)
at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
at scripts.IntelliFlow$FreeplaneTools$2.alternativeOpen(prototypeForIntelliflow.groovy:1938)
at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
at scripts.IntelliFlow$FreeplaneTools$2.run(prototypeForIntelliflow.groovy:1924)
at java.base/java.lang.Thread.run(Unknown Source)

**Reply by i-plasm** [reply]
Thanks, I fixed it: https://github.com/i-plasm/freeplane-scripts/blob/main/src/scripts/prototypeForIntelliflow.groovy

**Comment by nohaboat** [comment]
Thanks it is working now. Websearch opens the browser and searches the content. Node linked using node id takes to the requested node in the same map. Node linked using node's URI opens the requested node even in another mindmap.

File/Folder urls are not correctly identified and are opened in a browser window. But I can easily open them using freeplane keyboard shortcuts.

**Reply by i-plasm** [reply]
Hi @nohaboat , can you please clarify: What do you mean by "File/Folder urls are not correctly identified and are opened in a browser window"? Do you mean they are open in a web browser? Otherwise, do you mean they are parsed incorrectly therefore fail to open? If the latter, please give an example of a file url and a folder url that fails to be parsed.

**Reply by nohaboat** [reply]
Sorry for the late reply. Yes, I meant they open in web browser. I think it is difficult to distinguish between linux file urls (eg. /home/john/Documents/test.pdf) from search terms. Windows file urls don't have this problem because they start like C:\ . So what is happening is the file urls are being sent to web browser because the script thinks url is a search term that needs to be searched in a web browser. I have no problem with this limitation on linux because we have inbuilt freeplane commands to make file url hyperlinks and I use those to open files correctly.

Search function of IntelliFlow is great.
